ta/pkcs11/src/processing_symm.c

```diff
@@ -833,3 +833,77 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
 
 	return rc;
 }
+
+enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
+				     void *data, size_t data_sz,
+				     void **out_buf, size_t *out_sz)
+
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct active_processing *proc = session->processing;
+	void *in_buf = NULL;
+	size_t align = 0;
+	size_t in_sz = 0;
+
+	switch (proc->mecha_type) {
+	case PKCS11_CKM_AES_ECB:
+	case PKCS11_CKM_AES_CBC:
```

Comment:
Sorry for the mind flow below -- kept it in here in case it helps other reviewers too.

We do have [2.14 AES Key Wrap](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850517) which defines different values for mechanism as in here.

Wondering what kind of keys we should export with AES-ECB -- it is my understanding that ECB mode should not really be used for multiple blocks.

AES-CBC is slightly better.

If I look at [2.1.21 RSA AES KEY WRAP](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850425) this looks somewhat better approach.

Then I was interested in how the key material is then encoded before AES.

In above:
`The recommended format for an asymmetric target key being wrapped is as a PKCS8 PrivateKeyInfo`.
And this uses keyword `recommended` which is kinda a bit nice wording in standard -- but if I am not mistaken this would say DER encode the private key material.

Then I went back to look what these mechanism here then mean.

For AES-ECB (and for CBC) it says:
`For wrapping, the mechanism encrypts the value of the CKA_VALUE attribute of the key that is wrapped, padded on the trailing end with up to block size minus one null bytes so that the resulting length is a multiple of the block size. The output data is the same length as the padded input data. It does not wrap the key type, key length, or any other information about the key; the application must convey these separately.`

And this serialization is kinda clear.

But then what makes it slightly unclear -- we have source key material (eg. key to wrap) and then we have wrapping key material (eg. RSA, AES or so key that does encryption) and then wrapping mechanism (eg. how to wrap things).

So what defines what ;)

Wrapping mechanism is easy -- that comes as argument `pMechanism` to `C_WrapKey()`.

For `CKM_AES_ECB` and `CKM_AES_CBC` that is defined clearly I would say but that can only be used for secret key objects's `CKA_VALUE` field. I suppose it should disallow any other key objects.

So I am wondering that `get_key_data_to_wrap()` probably should be subject to mechanism and used wrapping key?

Then in example if we use `CKM_RSA_AES_KEY_WRAP` and want to wrap AES key then how should that then be encoded so that it would in example be compatible with something?

For me key wrapping use case would be to transfer key from source security world to target security world. And that would require either shared secret from target security world or public key from target security world. With RSA OAEP (or RSA PKCS1.5) I kinda understand then that using public key would be the best -- but what would be the real use case with these other wrapping methods. And then the better question -- what wrapping methods should be implemented?

Reply:
I am also a bit confused by how wrapping separates the encrypted secret value and its attributes. When referring to wrapped key blobs with HW support (on imx), didn't the key attributes were wrapped inside encrypted materials?

Reply:
Idea of this PR was more to add boiler plate code for Wrap/Unwrap using AES-ECB and CBC mechanism. See [TS-98](https://projects.linaro.org/browse/TS-98). I did look at the CKM_AES_KEY_WRAP/WRAP_PAD but that is dependent on RFC5649 and corresponding crypto implementation is not available in TEE core. So, I left it as it is.

For the question on what defines what - below is my understanding. We get wrapping key, key to be wrapped and mechanism in C_WrapKey()

AES_ECB/CBC/AES_KEY_WRAP - things are straightforward

For CKM_RSA_AES_KEY_WRAP ->
get_key_data_to_wrap() -> key data to be wrapped 
In my understanding this depends on the target key type which is being wrapped.

This key is wrapped by an AES key -> this is a temporary key. Based on the parameters passed with CKM_RSA_AES_KEY_WRAP,  required AES key of the provided size would be created. This key is used to wrap the key data above.

Finally the RSA private key is used to wrap the AES key. Again here only value of the AES key should get wrapped.

When unwrapping, the same mechanism and parameters are passed, so unwrapping happens in reverse direction.

In the complete flow, the data which is wrapped is not dependent on the mechanism. There have been publications  that this is weak and attack prone. It is worth wrapping attributes of the key along with key data. There has been proposal for the PKCS#11 but nothing in specs as yet.

I agree with the use case you have described. Here the boiler plate code is where you have a shared secret between 2 parties. We should eventually add a mechanism like CKM_RSA_PKCS/PKCS_OEAP to this list. These seem like the mechanisms mostly used. I tried searching for usecases of mechanism - CKM_RSA_AES_KEY_WRAP but couldn't find much.

@etienne-lms , my understanding is wrapping doesn't take into account the attributes of the key as per PKCS#11 specification. This is an issue but nothing in specification on it as yet.